Buevest german backtrans

BueVest opened the pull request:
Still preliminary work. G0 is about ready for testing in a wider context. G1 should be relatively easy to add.

BueVest commented:
Apparently, the Travis build fails with the following message.
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
163
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received

What could be going wrong?
Nothing seems to fail locally.

egli commented:
@BueVest I just restarted that particular test and now all is green. Might have been a fluke

BueVest commented:
I think we should merge this PR now. It contains fully working back-translatable grade 0 and 1. Grade 2 will require some substantial development and a lot of work. So, I think it would be better to do that in another PR. Then we can start using g0 and g1 and gain some experience.

bertfrees commented:
@BueVest OK, we'll see what we can do. It is a bit late that you tell this and there are a lot of changes to review. Tomorrow is release day.

bertfrees commented:
Had a quick look. Here are some comments:

- What's with the temporary test file?
- Why do we need separate "bd" tables? Couldn't both forward and backward translations be defined in the same tables?
- You have created quite a lot of duplication with the "bd" tables. Even if there is a need to have two sets of tables, could something be done about the duplication?
- Why do we need the "bd" tests?
- Changes were made to de-chardefs6.cti, but this was not reflected in the tests.

BueVest commented:
You wrote:

*	What's with the temporary test file?

Sorry, I thought I had already removed it. Didn’t I move it to the proper test directory?

 

*	Why do we need separate "bd" tables? Couldn't both forward and backward translations be defined in the same tables?

BD is for “bidirectional” (feel free to rename). So, they are in fact for both forward and backward translation. Some aspects of the Braille created by the original (non-BD) tables are less suited for back-translation, e.g. the lack of capital letters, the unconditional removal of some spaces, the use of the not so detailed accented letters etc. This corresponds to the difference between the Danish 6 dots grade 1 and 2 tables and their literary counterparts. The literary tables generate beautiful Braille for reading, but it is less suited for back-translation than the Braille created by the other tables. Hope it makes sense.

 

*	You have created quite a lot of duplication with the "bd" tables. Even if there is a need to have two sets of tables, could something be done about the duplication?

Yes, that would be nice indeed, but that would require a change of the file structure for the German tables. We could move the lines that can be shared into new files, which are then shared between the forward/literary tables and the bidirectional ones. However, if this change was to be made, I would rather make it in collaboration with the original table author, and not by myself. Currently, the chardefs file is shared, because I could make the necessary changes without affecting forward translation in the original tables.

 

*	Why do we need the "bd" tests?

The BD tests test the bidirectional tables in both directions. The other tests test the original tables, and only in the forward direction. The files could in theory be merged, but since they test two separate sets of tables, I think they should be separate yaml files.

 

*	Changes were made to de-chardefs6.cti, but this was not reflected in the tests.

The changes are primarily to the order of character definitions to ensure correct back-translation. They should not affect forward translation.

 

If there are any big problems with merging now, we can do it later. I just thought it would be a good idea to gain some broader user experience with the tables now. Especially, since people are currently using the “forward only” tables for back-translation with screen readers, e.g. NVDA, and flagging back-translation problems as bugs rather than request for new features. Perhaps we should flag it as an error to use a forward table for back-translation? Just thinking…

bertfrees commented:
> Some aspects of the Braille created by the original (non-BD) tables are less suited for back-translation, e.g. the lack of capital letters, the unconditional removal of some spaces, the use of the not so detailed accented letters etc.

Yes, I assumed it was like the Danish tables, but still: couldn't they be combined in the same table? Or would that be too confusing?

> We could move the lines that can be shared into new files, which are then shared between the forward/literary tables and the bidirectional ones. However, if this change was to be made, I would rather make it in collaboration with the original table author, and not by myself. Currently, the chardefs file is shared, because I could make the necessary changes without affecting forward translation in the original tables.

OK. I think it would be good to do this refactoring right from the start. This also gives us the opportunity to get some feedback from the original table author.

> The BD tests test the bidirectional tables in both directions. The other tests test the original tables, and only in the forward direction. The files could in theory be merged, but since they test two separate sets of tables, I think they should be separate yaml files.

I think it might be easier to follow if the tests were grouped into the things that work the same in both versions of the tables and the things were the tables differ.

> The changes [to de-chardefs6.cti] are primarily to the order of character definitions to ensure correct back-translation. They should not affect forward translation.

Hmm, are you sure? I thought I saw some actual differences?

> If there are any big problems with merging now, we can do it later. I just thought it would be a good idea to gain some broader user experience with the tables now.

Yes, it sure is nice to get user feedback as soon as possible, however we shouldn't use time pressure as an excuse to do things in a sloppy way. If you would have done the PR a week ago I'm sure we would have been able to fix all the issues. I do trust you that you would fix the issues afterwards if we would merge the PR now, but still...

> Perhaps we should flag it as an error to use a forward table for back-translation? Just thinking…

This is something that needs to be handled on the NVDA side ideally. Adding these kind of metadata based limitations into the library itself only works if you can select tables only through metadata, but this is not the case at the moment.

BueVest commented:
What needs to be done here?

Concerning Bert's suggestions above:
1. We could refactor, so that the "core" files contain all the common stuff, and the main tables contain the stuff specific to the individual tables. However, that would be some change to the original file structure, and I would rather not do that without the approval of the owner of the original tables. Is that you, @egli ?

2. Refactoring the yaml tests: @bertfrees , you apparently wanted to have the tests in fewer files and grouped to show the same things for the different tables. What is that supposed to look like? As far as I remember, we can run the same tests for multiple tables, but not with different flags for each table, e.g. forward for one table and both directions for another table, unless that has been changed in the code.

Any suggestions are welcome.

bertfrees commented:
@BueVest What happened in Git? There are two seemingly identical branches that got merged. Now the diff in Github has become unreadable. I cleaned it up locally and also rebased onto master. If you agree I can push it.

bertfrees commented:
>  You apparently wanted to have the tests in fewer files and grouped to show the same things for the different tables.

Yes. Well, it was just a suggestion. I think it might help in understanding how the tables differ. But I don't know how feasible it actually is. It depends on how much and what kind of differences there are. And you will indeed have to run the common tests in both directions for both tables.

Which brings us to one my others questions: Do you really need the two tables? Couldn't the backward behavior of your "bidi" table be combined with the forward behavior of the original table? Is the forward part of the "bidi" table really important to have? It's just a naive question. Maybe both tables are so different that they are not compatible at all? I don't know. I need to understand better how exactly the tables differ.

Let's assume for a moment the answer to the above question is yes (we do really need both directions of the new table). Couldn't the backward behaviors of the main and "bidi" tables be aligned? Some back-translation is always better than a back-translation that is not working at all, right?

bertfrees commented:
I think first and foremost we need a summary of things where the two tables differ.

BueVest commented:
Yes, the forward part is indeed important. It creates Braille that can be back-translated, whereas the Braille from the original tables can’t be back-translated. There are some distinct differences of which I have already mentioned the marking of capital letters and the processing of accented letters.

 

If the caller could set a flag, which could then be checked by various lines in the tables, then we could probably merge the two sets (#ifdef someFlag/ #ifndef someflag / #endif), but I don’t see how this could currently be done.

 

The whole thing about the different flavours of Braille within the same Braille code and why some flavours are more back-translatable than others is all about the different way we use Braille, i.e. reading books (possibly on paper) vs. reading and writing documents with a screen reader or on a Braille note-taker. It is a discussion which is probably appropriate to most languages, except where they have deliberately changed the Braille code to make back-translation easier, e.g. UEB.

 

If you are interested, I will be happy to try to explain it to you in more details, but it will probably be easier over Skype than through a PR.

bertfrees commented:
So what about the second part of my question: Could the backward behavior of the main table be the same as that from the bidi table?

Maybe I just don't know enough details, but from my vague understanding of it it sounds like both variants of the braille are not incompatible. One just contains more information than the other (like information of capitals) that make the back-translation better.

If you want to do a Skype call to discuss this PR, that is fine for me. However I think we need some written explanation of the table anyway. It's interesting for me, but I'm not the only person who needs to know. It's also and primarily the original author of the table (Christian's colleague), and other German braille people, that need to understand.

bertfrees commented:
Regarding the tests, I gave it another thought: for dictionary tests it doesn't matter that there are a lot of files with possibly duplication. They are not meant as documentation. But it would be nice to have a YAML file that explains the differences between the main German braille code and the variant optimized for back-translation.

BueVest commented:
> So what about the second part of my question: Could the backward behavior of the main table be the same as that from the bidi table?

> Maybe I just don't know enough details, but from my vague understanding of it it sounds like both variants of the braille are not incompatible. One just contains > more information than the other (like information of capitals) that make the back-translation better.

 

Yes, you are right. A great part of the work is simply making the tables produce Braille which contains as much information as possible while still following the rules.

I am not quite sure, but the original tables could probably be made to perform the same back-translation as the BD tables. In fact, they would probably perform better now than before anyway, because I have changed the order of character definitions, which are common to the two sets of tables.

On the other hand, as far as I remember, the original tables have some weird work-around to get rid of unwanted capsigns. That might have to be re-worked quite a bit for proper back-translation within the same table.

 

I could describe the differences in forward translation between the two table sets within the bd tables themselves., provided that back-translation is simply not defined for the original tables. However, describing the whole philosophy of why you would want two sets of tables? That is somewhat harder to describe in few words. I guess it is something, which is perhaps obvious to Braille users, but perhaps not so much to many others. That is why I suggested a Skype call. If I am to describe it, I need to understand what it is that is so difficult to understand about it, so to speak.

Also, such an explanation should probably not be hidden away in a specific table, but rather be in the manual in the section about back-translation. It could be relevant for work on back-translation in any language that has advanced grade 2 or grade 3, and where the Braille code is not specifically made for automatic back-translation.

Hope it makes sense.

bertfrees commented:
> I could describe the differences in forward translation between the two table sets within the bd tables themselves, provided that back-translation is simply not defined for the original tables.

Yes! That's exactly what I'm after.

> [...] describing the whole philosophy of why you would want two sets of tables? [...] If I am to describe it, I need to understand what it is that is so difficult to understand about it, so to speak.

It is not so difficult to understand at all. I'm just asking a lot of questions (sometimes deliberately naive) to make sure that we are doing it the best possible way. For example I think whether we should try to handle both variants of the braille code within the main table (or both tables) when back-translating is a valid question. If it is indeed doable it would be a major improvement. You wouldn't need to figure out which table to select, you can just pick one and it would work.

Anyway, I'm not asking you to describe the whole philosophy of the two sets of table. I think it would indeed be a valuable addition, and useful for developing other tables, but for now all I'm asking for is a proper description of the behavior.

bertfrees commented:
By the way can I push the cleaned up branch that I have locally or not? I want to be able to look at the combined diff on Github, that is currently not possible.

BueVest commented:
Yes, of course you can.

bertfrees commented:
OK, pushed.

bertfrees commented on tables/de-g0-bd.utb:
```diff
@@ -0,0 +1,91 @@
+# - Gross-/Kleinschreibung wird nicht berücksichtigt (d.h. für
+#   Abkürzungen u.ä. braucht es Vorprogramm)
+# - Akzente: Hinweis auf die beiden Tabellen (bis jetzt nur als
+#   Kommentar in de-g0.utb)
```
This text needs to be corrected.

bertfrees replied:
> Hmm, yes, I did consider it, but what does it actually mean, and what should we say instead, or should I just remove it?

Christian will do it.

bertfrees replied:
@egli

egli replied:
> Gross-/Kleinschreibung wird nicht berücksichtigt (d.h. für Abkürzungen u.ä. braucht es Vorprogramm)

Capitalization is ignored, i.e. for abbrevs you will need a pre-processor

> Akzente: Hinweis auf die beiden Tabellen (bis jetzt nur als Kommentar in de-g0.utb)

Not sure about this one. I think there is some special treatment of accents and maybe this is only documented in `de-g0.utb`?

bertfrees replied:
@egli You marked this resolved but it's not fixed.

egli replied:
I think I fixed this in 7259d41c4986785f599b653be2a3f2b309f77340

bertfrees replied:
OK I guess it was not pushed yet.

bertfrees commented on tables/de-g0-bd.utb:
```diff
@@ -0,0 +1,91 @@
+# -----------
+
+#-name: Deutsche Basisschrift
+#-display-name: German uncontracted braille
```
This metadata needs to be updated.

bertfrees replied:
> Then, what do you want instead? Should I just add (bidi)?

I'm not sure what it should be. You tell me :-) No, just adding "bidi" is way to criptic. It needs to be something understandable. Maybe "German uncontracted braille, back-translatable"?

bertfrees replied:
> How about simply “bidirectional”, which is what “bidi” stands for anyway?
>
> I think we need some “Liblouis terms” to describe these things properly. With the Danish tables, the bidirectional ones came first, so “lit” was appended to the other ones, which is also a bit cryptic.

Between "back-translatable" and "bidirectional", I find the former more to the point and it's more understandable language. But both mentioned options focus on the transcription process rather than the writing system itself. Maybe that is inevitable.

Yes, we need a good system-wide term for this category of tables. For Danish, the tables with "lit" and without "lit" actually have exactly the same display name at the moment, which is not good. I don't care about the file names of the tables, they are computer things so are allowed to be a bit cryptic. But the stuff that is visible to the user, like the `display-name` and `index-name` metadata fields, should be as clear as possible, and consistent across tables.

bertfrees replied:
I updated the metadata. I haven't changed the metadata of the Danish tables yet.

bertfrees replied:
It's too late to change it now, but I think the term "detailed" summarizes well what this braille system variant is.

bertfrees commented on tables/de-g0-bd.utb:
```diff
@@ -0,0 +1,91 @@
+#-------------------------------------------------------------------------------
+
+# must include display table directly to establish Character/dots associations.
+include de-eurobrl6.dis
```
This is a bit fishy.

bertfrees replied:
> Then, what do you want? Some display statements need to be there to define the dot patterns. The characters themselves have other definitions in a literary (none-computer Braille) table.

I think I would prefer if the display rules that are actually needed are inserted here instead of the include. But more importantly I'd like to understand the issue better. Which are the dot patterns that need to be defined and why?

bertfrees replied:
Changed the comment

BueVest commented:
Should I attach a note to the original g0-2 tables saying something about them being literary tables and not meant for back-translation? I can see that at least NVDA still selects them for input, i.e. back-translation.

bertfrees commented on tables/de-chardefs6.cti:
```diff
@@ -1,6 +1,7 @@
-# liblouis: Character definitions for german tables
+# liblouis: Character definitions for german tables (bidirectional)
```
Are we sure that the changes to de-chardefs6.cti don't affect the main table?

bertfrees replied:
> Yes. All tests pass, or did until we rebased.

Yes, I'm sure the tests pass, but the tests aren't that great. We only have a dictionary test. That means only words are tested. Anything special, like punctuation, is not covered.

I'm just looking at the file and while a lot of the changes are things that you just moved around, or where you added a `noback`, there also seem to be some actual changes in the dot patterns.

bertfrees replied:
> Ok, I actually just found something:
>
> With the original tables without any additional display table, accented letters now translate as ⠈letter, not “letter, i.e. ⠈a, not “a. This is an example of a dot pattern that needs to be mapped by an included display file, which I apparently forgot to do with the original tables.
>
> It actually should have been caught by the dictionary test, which contains words with accented letters. However, the yaml test file specifies its own display file and thereby masks the problem.
>
> I should include the display file in the original tables, same as in the bidi tables.

Nope. Wrong thinking :-) It is not a problem in the translation table. It is a problem in how you provide the braille input for a translation. The YAML test does it correctly: the braille encoding in the test matches the specified display table. You need to do the same.

bertfrees replied:
> Then, what solution would you suggest? Please, suggest something concrete instead of just saying “no”.

It's very simple. It's a matter of encoding/decoding braille correctly when interfacing with Liblouis. I'm not sure how you are invoking Liblouis, but if you either specify a display table and encode/decode your braille according to that table, or if you use the dotsIO mode, and you still get an unexpected result, _then_ there is a problem. Otherwise there is no problem.

bertfrees replied:
> Yes, I agree, and all is good and well, as long as the caller specifies a display table, which, unfortunately, they don’t at the moment. I know for certain that both the HIMS note-takers and NVDA call liblouis without specifying a display table, and I am almost sure about JAWS, even though I don’t have access to the source code. Personally, I test these things with lou_trace, with or without a display table.
>
> For a long term solution, I would suggest that the caller should be required to specify a display table, or that the table can specify a “default display table” to use with that table. No, this is not mixing the display opcode back with the others. It is simply a way for the table author to specify the default Braille representation for the table. This is typically guided by nationality and language. For Danish it would be da-dk-octobraille.dis. For German, it would typically be de-eurobrl6.dis.
>
> Yes, I am also looking forward to the day when we all use Unicode Braille for our embossers and notetakers, but, unfortunately, that is not the present situation.
>
> For a short term solution, I still suggest we include de-eurobrl6.dis in the German tables, even if it is “wrong thinking”. Then, they still work without specifying a display table explicitly, which is unfortunately required at the moment. If you have a better practical solution for now, please, tell me about it.

My solution is to force programs that use Liblouis to specify a display table. The reality is that if we do nothing, HIMS, NVDA, etc. are forever going to keep doing it like they do it today. It is just laziness from their part. There is no reason to wait for that. Sometimes you just have to make blunt decisions to get something going.

I don't even understand how a program like NVDA can get away with what they do (if it is indeed like you say)? Let's for a minute assume that you can be sure that everyone who uses the Danish Liblouis table has their braille device configured to use the character set that matches exactly octobraille.dis (which is hard to believe, but OK). What do you do if you want to read German text with NVDA and for that you switch to the German table? Do you change the settings of your braille display accordingly? That would be terrible!

But OK, you convinced me, we can include de-eurobrl6.dis in the German table. As long as it's allowed it is indeed better to not unnecessarily break the current behavior. Before we stop allowing using a translation tables as display table, we should make sure we replace or convert them with proper display tables anyway, and by including de-eurobrl6.dis we already know what the replace it with when the time comes.

By the way it is probably also a good idea to start add metadata to the display tables. It will become useful.

bertfrees replied:
> Luckily, the situation is not quite that bad when using a screen reader with a display. You don’t configure the character/display table on the display, but only on the computer or tablet/phone. I am not quite sure about the protocol for communication between a screen reader and a Braille display, but it seems to be communicating dots rather than their character equivalents. It is an old serial protocol, which probably hasn’t changed since the first Braille displays in the 90s.
>
> However, the situation with embossers and note-takers is more or less like you describe it. If I receive a .brl file from a German, I will have to change my note-taker or embosser to use the correct table before trying to read or emboss the file. And yes, all embossers in Denmark are set up to use the OctoBraille table, unless you come across a really old installation that was set up to use US computer Braille. This is why I am looking forward to all of us just using Unicode Braille and nothing else.

So if we're just talking about screen readers, they _must_ be specifying a display table or using the dotsIO mode, right? Otherwise they have no idea which dots they are sending to the braille display, and the protocol apparently requires it.

Even if the situation with embossers and note-takers is like it is, my standpoint is still that any software that makes use of Liblouis, if it can update to new versions of Liblouis, can also be changed to specify display tables.

bertfrees commented on tables/de-g1-bd-core.cti:
```diff
@@ -0,0 +1,44 @@
+# liblouis: German grade 1 braille (bidirectional)
```
This table should be merged with de-g1-core.cti

bertfrees replied:
> I agree, but the lines that are different between the two tables would then have to be moved to the main tables. This is a change to the structure of the original tables, so I think I must have the approval of the owner before doing this refactoring.

You don't need approval of the owner to do structural changes. Especially not in the PR phase.

But anyway, the only lines that are different are the `nofor` lines that you added. So at the very least you can include de-g1-core.cti in de-g1-bd-core.cti. But you should also just be able to add the `nofor` lines to de-g1-core.cti because they won't affect the forward translation.

> Actually, I originally also made the Danish tables that way, with common core files, but I was then asked to split the tables and keep them separate, even though they share most lines, like the German tables.

Who asked you that?

With the Danish tables it's slightly less of an issue I think because all the files are maintained by you anyway, and we know it is done well. But in this case, you are either asking another person to suddenly maintain a lot more files, or you maintain the new table yourself but that means they can easily diverge if no precautions are made. 

> I suggest we keep them separate for now and then make it a milestone for the next release, perhaps both for the German and the Danish Tables.

OK, I can live with that.

bertfrees replied:
See https://github.com/liblouis/liblouis/issues/911

bertfrees commented on tables/de-g1-bd-core.cti:
```diff
@@ -0,0 +1,44 @@
+
+include de-g1-core-patterns.dic
+
+noback nocross    au         16                           # 1 #
```
`nocross` should be changed to `nocross always` (this is because I have rebased the branch)

bertfrees replied:
Done

bertfrees commented on tables/de-g1-bd.ctb:
```diff
@@ -0,0 +1,85 @@
+# - Gross-/Kleinschreibung wird nicht berücksichtigt (d.h. für
+#   Abkürzungen u.ä. braucht es Vorprogramm)
+# - Akzente: Hinweis auf die beiden Tabellen (bis jetzt nur als
+#   Kommentar in de-g0.utb)
```
This text needs to be updated.

bertfrees replied:
@egli

egli replied:
isn't this the same as above?

bertfrees replied:
Probably yes.

bertfrees commented on tables/de-g1-bd.ctb:
```diff
@@ -0,0 +1,85 @@
+# -----------
+#-name: Deutsche Vollschrift
+#-index-name: German, partially contracted
+#-display-name: German partially contracted braille
```
Update metadata

bertfrees replied:
Done

bertfrees commented on tables/de-accents-detailed.cti:
```diff
@@ -1,6 +1,7 @@
 # liblouis: Letters with accents (without ä ö ü) - detailed representation of french and italian accents
 #
 #  Copyright (C) 2018 SBS Schweizerische Bibliothek für Blinde, Seh- und Lesebehinderte
+#  Copyright (C) 2020 Bue Vester-Andersen
```
A copyright line is not really appropriate in case of an addition of a single comment.

bertfrees replied:
> Should there be anything else to indicate that I made changes to the file?

No, that's not needed.

bertfrees commented on tables/de-g0-bd-core.uti:
```diff
@@ -0,0 +1,810 @@
+# liblouis: German grade 0 braille (bidirectional)
```
We should try to remove the duplication between this file and de-g0-core.uti.

bertfrees replied:
> Agree, but let us make it an item for the next release.

OK then.

bertfrees replied:
See https://github.com/liblouis/liblouis/issues/911

bertfrees commented on tables/Makefile.am:
```diff
@@ -71,9 +71,13 @@ table_files = \
 	de-g0-core.uti \
 	de-g0.utb \
+	de-g0-bd-core.uti \
+	de-g0-bd.utb \
```
"bidi" makes more sense as an abbreviation of "bidirectional" than "bd" so I think I will renamed the files. But you don't have to worry about that Bue.

bertfrees replied:
Done

bertfrees commented:
> Should I attach a note to the original g0-2 tables saying something about them being literary tables and not meant for back-translation?

OK makes sense.

BueVest commented:
> +# - Gross-/Kleinschreibung wird nicht berücksichtigt (d.h. für
+#   Abkürzungen u.ä. braucht es Vorprogramm)
+# - Akzente: Hinweis auf die beiden Tabellen (bis jetzt nur als
+#   Kommentar in de-g0.utb)

This text needs to be corrected.

 

Hmm, yes, I did consider it, but what does it actually mean, and what should we say instead, or should I just remove it?

 

> +#-name: Deutsche Basisschrift
+#-display-name: German uncontracted braille

This metadata needs to be updated.

Then, what do you want instead? Should I just add (bidi)?

 

In  <https://github.com/liblouis/liblouis/pull/842#discussion_r385662450> tables/de-g0-bd.utb:

> +# must include display table directly to establish Character/dots associations.
+include de-eurobrl6.dis

This is a bit fishy.

 

Then, what do you want? Some display statements need to be there to define the dot patterns. The characters themselves have other definitions in a literary (none-computer Braille) table.

 

In  <https://github.com/liblouis/liblouis/pull/842#discussion_r385664038> tables/de-chardefs6.cti:

> -# liblouis: Character definitions for german tables
+# liblouis: Character definitions for german tables (bidirectional)

Are we sure that the changes to de-chardefs6.cti don't affect the main table?

Yes. All tests pass, or did until we rebased.

 

In  <https://github.com/liblouis/liblouis/pull/842#discussion_r385665034> tables/de-g1-bd-core.cti:

> @@ -0,0 +1,44 @@
+# liblouis: German grade 1 braille (bidirectional)

This table should be merged with de-g1-core.cti

 

I agree, but the lines that are different between the two tables would then have to be moved to the main tables. This is a change to the structure of the original tables, so I think I must have the approval of the owner before doing this refactoring.

Actually, I originally also made the Danish tables that way, with common core files, but I was then asked to split the tables and keep them separate, even though they share most lines, like the German tables.

I suggest we keep them separate for now and then make it a milestone for the next release, perhaps both for the German and the Danish Tables.

 

+noback nocross    au         16                           # 1 #

nocross should be changed to nocross always (this is because I have rebased the branch)

 

No problem.

 

 

#  Copyright (C) 2018 SBS Schweizerische Bibliothek für Blinde, Seh- und Lesebehinderte
+#  Copyright (C) 2020 Bue Vester-Andersen

A copyright line is not really appropriate in case of an addition of a single comment.

No problem, I will remove it. Should there be anything else to indicate that I made changes to the file?

 

In  <https://github.com/liblouis/liblouis/pull/842#discussion_r385667362> tables/de-g0-bd-core.uti:

> @@ -0,0 +1,810 @@
 
+# liblouis: German grade 0 braille (bidirectional)
 

We should try to remove the duplication between this file and de-g0-core.uti.

Agree, but let us make it an item for the next release.

 

In tables/Makefile.am <https://github.com/liblouis/liblouis/pull/842#discussion_r385667890> :

> @@ -71,9 +71,13 @@ table_files = \
 
        de-eurobrl6u.dis \
 
        de-g0-core.uti \
 
        de-g0.utb \
 
+       de-g0-bd-core.uti \
 
+       de-g0-bd.utb \
 

"bidi" makes more sense as an abbreviation of "bidirectional" than "bd" so I think I will renamed the files. But you don't have to worry about that Bue.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/liblouis/liblouis/pull/842?email_source=notifications&email_token=ADAVTCOISKVGCOJRWHDNQB3RFD7GLA5CNFSM4IT5GVXKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCXLBNFQ#pullrequestreview-366352022> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADAVTCOVQILAABQ2JBX7JZDRFD7GLANCNFSM4IT5GVXA> .

BueVest commented:
I'm not sure what it should be. You tell me :-) No, just adding "bidi" is way to criptic. It needs to be something understandable. Maybe "German uncontracted braille, back-translatable"?

 

How about simply “bidirectional”, which is what “bidi” stands for anyway?

 

I think we need some “Liblouis terms” to describe these things properly. With the Danish tables, the bidirectional ones came first, so “lit” was appended to the other ones, which is also a bit cryptic.

BueVest commented:
In  <https://github.com/liblouis/liblouis/pull/842#discussion_r385727034> tables/de-chardefs6.cti:

> -# liblouis: Character definitions for german tables
+# liblouis: Character definitions for german tables (bidirectional)

Yes. All tests pass, or did until we rebased.

Yes, I'm sure the tests pass, but the tests aren't that great. We only have a dictionary test. That means only words are tested. Anything special, like punctuation, is not covered.

I'm just looking at the file and while a lot of the changes are things that you just moved around, or where you added a noback, there also seem to be some actual changes in the dot patterns.

 

I have changed some characters from their computer Braille representation to their actual literary representation where I could do this without disrupting the original tables. This was necessary to provide backtranslation for these characters.

If I am to merge the tables and change the structure, I will have to make proper tests for the original tables anyway to make sure I don’t break anything in the process. I will see to it before the next release.

BueVest commented:
Yes, I'm sure the tests pass, but the tests aren't that great. We only have a dictionary test. That means only words are tested. Anything special, like punctuation, is not covered.

I'm just looking at the file and while a lot of the changes are things that you just moved around, or where you added a noback, there also seem to be some actual changes in the dot patterns.

 

Ok, I actually just found something:

With the original tables without any additional display table, accented letters now translate as ⠈letter, not “letter, i.e. ⠈a, not “a. This is an example of a dot pattern that needs to be mapped by an included display file, which I apparently forgot to do with the original tables.

 

It actually should have been caught by the dictionary test, which contains words with accented letters. However, the yaml test file specifies its own display file and thereby masks the problem.

 

I should include the display file in the original tables, same as in the bidi tables.

BueVest commented:
Nope. Wrong thinking :-) It is not a problem in the translation table. It is a problem in how you provide the braille input for a translation. The YAML test does it correctly: the braille encoding in the test matches the specified display table. You need to do the same.

 

Then, what solution would you suggest? Please, suggest something concrete instead of just saying “no”.

bertfrees commented:
I've added this to the 3.13 milestone because I believe we can solve the remaining issues, apart from the refactoring stuff and the addition of tests, but we agreed to tackle those next time.

BueVest commented:
It's very simple. It's a matter of encoding/decoding braille correctly when interfacing with Liblouis. I'm not sure how you are invoking Liblouis, but if you either specify a display table and encode/decode your braille according to that table, or if you use the dotsIO mode, and you still get an unexpected result, then there is a problem. Otherwise there is no problem.

 

Yes, I agree, and all is good and well, as long as the caller specifies a display table, which, unfortunately, they don’t at the moment. I know for certain that both the HIMS note-takers and NVDA call liblouis without specifying a display table, and I am almost sure about JAWS, even though I don’t have access to the source code. Personally, I test these things with lou_trace, with or without a display table.

 

For a long term solution, I would suggest that the caller should be required to specify a display table, or that the table can specify a “default display table” to use with that table. No, this is not mixing the display opcode back with the others. It is simply a way for the table author to specify the default Braille representation for the table. This is typically guided by nationality and language. For Danish it would be da-dk-octobraille.dis. For German, it would typically be de-eurobrl6.dis.

Yes, I am also looking forward to the day when we all use Unicode Braille for our embossers and notetakers, but, unfortunately, that is not the present situation.

 

For a short term solution, I still suggest we include de-eurobrl6.dis in the German tables, even if it is “wrong thinking”. Then, they still work without specifying a display table explicitly, which is unfortunately required at the moment. If you have a better practical solution for now, please, tell me about it.

BueVest commented:
So if we're just talking about screen readers, they must be specifying a display table or using the dotsIO mode, right? Otherwise they have no idea which dots they are sending to the braille display, and the protocol apparently requires it.

 

Yes, perhaps. I am not sure, but it would be interesting to look into at some time.

 

Even if the situation with embossers and note-takers is like it is, my standpoint is still that any software that makes use of Liblouis, if it can update to new versions of Liblouis, can also be changed to specify display tables.

 

Yes, of course they can, but we should then do the following:

1.	Announce the change of policy clearly, both in the NEWS file and the manual.
2.	Flag it as an error or at least a warning, if a display table is not specified.
3.	Alternatively, let the table author specify a default display table to be used, either as an opcode or as metadata, which can be read and used by Liblouis itself.

 

BTW: Is there anything more for me to do in this PR?

bertfrees commented:
> Alternatively, let the table author specify a default display table to be used, either as an opcode or as metadata, which can be read and used by Liblouis itself.

Nice try.

> BTW: Is there anything more for me to do in this PR?

No, I can't think of anything. It is almost ready and I we should be able to handle the remaining things ourselves. I've created an issue for the next release, you probably saw that.

bertfrees commented:
You should probably read the whole discussion. We talked about all of these things.
